test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestScenariosCrossProduct.java

```diff
@@ -0,0 +1,114 @@
+ */
+
+public class TestScenariosCrossProduct {
+    public static void main(String[] args) {
+        // Single set should test all flags in the set by themselves.
+        try {
+            TestFramework t1 = new TestFramework();
+            t1.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=51",
+                                               "-XX:TLABRefillWasteFraction=53",
+                                               "-XX:TLABRefillWasteFraction=64"));
+            t1.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!e.getMessage().contains("The following scenarios have failed: #0, #1, #2")) {
+                throw e;
+            }
+        }
+
+        // The cross product of a set with one element and a set with three elements is three sets.
+        try {
+            TestFramework t2 = new TestFramework();
+            t2.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53"),
+                                        Set.of("-XX:+UseNewCode", "-XX:+UseNewCode2", "-XX:+UseNewCode3"));
+            t2.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!e.getMessage().contains("The following scenarios have failed: #0, #1, #2")||
+                 e.getMessage().contains("The following scenarios have failed: #0, #1, #2, #3")) {
+                throw e;
+            }
+        }
+
+        // The cross product of two sets with two elements is four sets.
+        try {
+            TestFramework t3 = new TestFramework();
+            t3.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53", "-XX:TLABRefillWasteFraction=64"),
+                                        Set.of("-XX:+UseNewCode", "-XX:-UseNewCode"));
+            t3.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!e.getMessage().contains("The following scenarios have failed: #0, #1, #2, #3") ||
+                 e.getMessage().contains("The following scenarios have failed: #0, #1, #2, #3, #4")) {
+                throw e;
+            }
+        }
+
+        try {
+            TestFramework t4 = new TestFramework();
+            t4.addCrossProductScenarios(Set.of("-XX:TLABRefillWastFraction=50 -XX:+UseNewCode", "-XX:TLABRefillWasteFraction=40"),
+                                        Set.of("-XX:+UseNewCode2"));
+            t4.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!(e.getMessage().contains("The following scenarios have failed: #0") ||
+                  e.getMessage().contains("The following scenarios have failed: #1")) ||
+                 e.getMessage().contains("The following scenarios have failed: #0, #1")) {
+                throw e;
+            }
+        }
+    }
```

Comment:
Could we also have a case with empty strings? Maybe it is already possible.
That would allow us to keep things "default". We don't always know what that is.
Imagine we may want a runt with the default value for `-XX:TLABRefillWastFraction=...` but that value could be different on different platforms.
Does that make sense?